include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -42,21 +44,62 @@ struct DefaultKernelName;
 template <typename KernelName = DefaultKernelName>
 class device_policy
 {
+    // Needed for the copy constructor that rebinds the kernel name
+    template <typename>
+    friend class device_policy;
+
   public:
     using kernel_name = KernelName;
 
     device_policy() = default;
+
     template <typename OtherName>
-    device_policy(const device_policy<OtherName>& other) : q(other.queue())
+    device_policy(const device_policy<OtherName>& other)
+    {
+        ::std::scoped_lock lock{other.mtx};
+        q = other.q;
+    }
+
+    device_policy(const device_policy& other)
+    {
+        ::std::scoped_lock lock{other.mtx};
+        q = other.q;
+    }
+
+    device_policy(device_policy&& other)
     {
+        ::std::scoped_lock lock{other.mtx};
+        q.swap(other.q);
     }
+
+    device_policy&
+    operator=(const device_policy& other)
+    {
+        ::std::scoped_lock lock{mtx, other.mtx};
```

Comment:
I believe it's not done. What if it's self assignment? How does `scoped_lock` behave for the same mutex passed twice? My bet it's a deadlock (better to say, lifelock)

Reply:
Yes, @rarutyun, you are right. Agreed.

Reply:
I modified the code to now check for self assignment.